ocpp-j-api/src/main/scala/com/thenewmotion/ocpp/json/api/client/SimpleClientWebSocketComponent.scala

```diff
@@ -53,44 +84,67 @@ trait SimpleClientWebSocketComponent extends WebSocketComponent {
         }
       }
 
-      def onError(ex: Exception) = {
-        logger.debug("Received error {}", ex)
-        SimpleClientWebSocketComponent.this.onError(ex)
+      def onError(ex: Exception): Unit = {
+        logger.debug("Received WebSocket error {}", ex)
+        if (connectedPromise.isCompleted)
+          SimpleClientWebSocketComponent.this.onError(ex)
+        else {
+          connectedPromise.complete(Failure(ex))
```

Comment:
I think race condition is possible here. this should be safer (and shorter):
```scala
if (!connectedPromise.tryComplete(Failure(ex))) {
  SimpleClientWebSocketComponent.this.onError(ex)
}
```

Reply:
Oopsie, indeed. I was hoping that java-websocket would make sure that event handlers are not run in parallel, but it's better to be sure than to hope :)